src/Body.ts

```diff
@@ -1,44 +1,590 @@
+			let cleared = true;
+			let addedKeys: string[] = [];
+			const keyPatches: { [ index: number]: string[] } = {};
+			let previousKeyIndex = 0;
+			for (const currentKey of currentKeys) {
+				const foundAtIndex = previousKeys.indexOf(currentKey, previousKeyIndex);
+				if (foundAtIndex === -1) {
+					addedKeys.push(currentKey);
+				}
+				else {
+					if (addedKeys.length) {
+						keyPatches[previousKeyIndex] = addedKeys;
+						addedKeys = [];
+					}
+
+					cleared = false;
+					previousKeyIndex = (foundAtIndex + 1);
+				}
+			}
+			if (addedKeys.length) {
+				keyPatches[previousKeys.length] = addedKeys;
+			}
+
+			// If all keys are new, we can start from scratch
+			if (cleared) {
+				itemElementMap.clear();
+				return this.render();
+			}
+
+			// Use previous keys to watch for deleted items
+			// and insert new keys at the indexes detected above
+			for (let i = 0, il = previousKeys.length; i <= il; i++) {
+				const key = previousKeys[i];
+
+				const keyPatch = keyPatches[i];
+				if (keyPatch) {
+					for (const addedKey of keyPatch) {
+						// Insert any newly introduced items
+						// that were added at this index
+						children.push(itemsByKey[addedKey]);
+
+						// Add to the updated element map
+						const update = itemElementMap.get(addedKey);
+						if (update) {
+							updatedElementMap.set(addedKey, update);
+						}
+					}
+				}
+
+				if (i < il) {
+					const item = itemsByKey[key];
+					const details = itemElementMap.get(key);
+					if (item) {
+						// This item has neither been added nor removed
+						// since the last render
+						children.push(item);
+
+						// Add to the updated element map
+						const update = itemElementMap.get(key);
+						if (update) {
+							updatedElementMap.set(key, update);
+						}
+					}
+					else if (details) {
+						// This item was deleted since the last render
+						if (!details.remove && details.element) {
+							// Store its rendered height before it is removed from DOM
+							// as it will not be added as a row in the next render
+							details.height = details.element.offsetHeight;
+						}
+						// Mark this item as having been deleted
+						details.remove = true;
+
+						// Add to the updated element map.
+						// This entry will be deleted once its size
+						// is taken into account in onElementUpdated
+						const updated = itemElementMap.get(key);
+						if (updated) {
+							updatedElementMap.set(key, updated);
+						}
+					}
+				}
+			}
+
+			// Store the updated item map
+			this._itemElementMap = updatedElementMap;
+		}
+
+		// Create a bottom margin if the data doesn't extend all the way to the end
+		if (start + items.length < max) {
+			children.push(v('div', {
+				key: 'marginBottom',
+				styles: {
+					height: ('10000px')
+				}
+			}));
+		}
+
+		return v('div', {
+			classes: this.classes(css.scroller),
+			key: 'scroller',
+			onscroll: this.onScroll
+		}, children);
+	}
+
+	protected visibleKeys() {
```

Comment:
Might be better if this is a property with `get` defined, or rename the method to `getVisibleKeys`